exonum-java-binding/core/rust/src/storage/temporarydb.rs

```diff
@@ -72,23 +72,20 @@ pub extern "system" fn Java_com_exonum_binding_core_storage_database_MemoryDb_na
     utils::unwrap_exc_or_default(&env, res)
 }
 
-/// Merges the given fork into the database.
+/// Merges the given patch into the database.
+/// The provided `patch_handle` is invalidated after the procedure and the
+/// Rust side is responsible for it.
 #[no_mangle]
 pub extern "system" fn Java_com_exonum_binding_core_storage_database_MemoryDb_nativeMerge(
     env: JNIEnv,
     _: JObject,
     db_handle: Handle,
-    view_handle: Handle,
+    patch_handle: Handle,
 ) {
     let res = panic::catch_unwind(|| {
-        let _db = handle::cast_handle::<TemporaryDB>(db_handle);
-        let _fork = match *handle::cast_handle::<View>(view_handle).get() {
-            ViewRef::Snapshot(_) => panic!("Attempt to merge snapshot instead of fork."),
-            ViewRef::Fork(fork) => fork,
-        };
-        //FIXME: Implement merging via db.merge(fork.into_patch())
-        //        db.merge(fork.patch().clone())
-        //            .expect("Unable to merge fork");
+        let db = handle::cast_handle::<TemporaryDB>(db_handle);
+        let patch = handle::acquire_handle_ownership::<Patch>(patch_handle);
+        db.merge(*patch).expect("Unable to merge fork");
```

Comment:
Will the context from the error be included into the exception message? In which realistic cases can it fail?

Reply:
The error message will be included, as well as stacktrace on debug builds. It can fail because of some internal RocksDB error, so we can't prevent and recover from this.